lib/datadog/core/telemetry/collector.rb

```diff
@@ -0,0 +1,218 @@
+        end
+
+        def appsec
+          version = appsec_version
+          Telemetry::V1::AppSec.new(version: version) if version
+        end
+
+        def appsec_version
+          tracer_version if Datadog.configuration.respond_to?(:appsec) && Datadog.configuration.appsec.enabled
+        end
+
+        def configuration_variables
+          configurations = []
+          environment_collector = Core::Diagnostics::EnvironmentCollector.new
+          [
+            Telemetry::V1::Configuration.new(name: 'diagnostics.debug', value: environment_collector.debug),
+            Telemetry::V1::Configuration.new(name: 'diagnostics.health_metrics.enabled',
+                                             value: environment_collector.health_metrics_enabled),
+            Telemetry::V1::Configuration.new(name: 'diagnostics.startup_logs.enabled',
+                                             value: Datadog.configuration.diagnostics.startup_logs.enabled),
+            Telemetry::V1::Configuration.new(name: 'profiling.advanced.endpoint.collection.enabled',
+                                             value: Datadog.configuration.profiling.advanced.endpoint.collection.enabled),
+            Telemetry::V1::Configuration.new(name: 'profiling.advanced.code_provenance_enabled',
+                                             value: Datadog.configuration.profiling.advanced.code_provenance_enabled),
+            Telemetry::V1::Configuration.new(name: 'profiling.advanced.legacy_transport_enabled',
+                                             value: Datadog.configuration.profiling.advanced.legacy_transport_enabled),
+            Telemetry::V1::Configuration.new(name: 'runtime_metrics.enabled',
+                                             value: environment_collector.runtime_metrics_enabled),
+            Telemetry::V1::Configuration.new(name: 'tracing.analytics.enabled',
+                                             value: environment_collector.analytics_enabled),
+            Telemetry::V1::Configuration.new(name: 'tracing.enabled', value: environment_collector.enabled),
+            Telemetry::V1::Configuration.new(name: 'tracing.priority_sampling',
+                                             value: environment_collector.priority_sampling_enabled),
+            Telemetry::V1::Configuration.new(name: 'tracing.partial_flush.enabled',
+                                             value: environment_collector.partial_flushing_enabled),
+            Telemetry::V1::Configuration.new(name: 'tracing.sampler.rate', value: environment_collector.sample_rate),
+            Telemetry::V1::Configuration.new(name: 'tracing.sampler.rules', value: environment_collector.sampling_rules),
+            Telemetry::V1::Configuration.new(name: 'agent_url', value: environment_collector.agent_url),
+            Telemetry::V1::Configuration.new(name: 'tags', value: environment_collector.tags)
+          ].each do |configuration|
```

Comment:
Hey! I haven't been following along the discussions so pardon me if this was already brought up/discussed elsewhere but this approach seems a bit... duplicate.

E.g. in here we're duplicating all of the names of a bunch of configuration settings, as well as needing to have _one more place_ we need to remember to update when adding new configuration settings.

Have you considered automatically probing the existing configuration and exposing what is there?

Since telemetry is part of the core, we could even, if needed, modify the settings DSL we use to opt in/out of sending fields, as needed.

Reply:
This is actually a little trickier than first thought....

I agree that repetition isn't as flexible as something that's discoverable. It may be convenient to simply discover what exists. There are trade-offs though.

1. Dynamically generated configuration means scaling. We don't want to send all fields (for cost and relevance reasons), so it necessitates more metadata.
2. Necessarily, you'd have to add metadata to every configuration setting, describing whether to keep or send the field.

Making configuration know about telemetry seems outside the scope of configuration. This I don't like and would argue against.

There are some things we could do within the scope of configuration behavior that may be helpful, for example, converting the configuration object to JSON or YAML. From that you could read the "path" & value that you need. But it doesn't solve the whitelist/blacklist behavior.

Personally, I think it's totally permissible, even desirable, to have an enumerated list of keys we want to explicitly capture. e.g. something like:

```ruby
# Whitelist of desired options
TARGET_OPTIONS = [
  'tracing.priority_sampling.enabled'.freeze,
  'tracing.partial_flush.enabled'.freeze,
  # ...
].freeze

# Returns { "path.to.option", "value" }
config_options = Datadog.configuration.to_hash
config_options_to_keep = {}

# Filter options
TARGET_OPTIONS.each do |option|
  config_options_to_keep[option] = config_options[option] if config_options.key?(option)
end

config_options_to_keep
```

Reply:
Also we shouldn't use the environment collector... it's not a utility/API, but a feature. Layering features on top of each other like this makes it harder to refactor them. And the environment collector is in need of refactoring.

All the settings you need should be in the settings object.